### PR TITLE
Add queue interop driver

### PR DESCRIPTION
**Pull request by makasim**
The driver can work with any [queue interop](https://github.com/queue-interop/queue-interop) compatible transport such as:

* [enqueue/amqp-ext](https://github.com/php-enqueue/amqp-ext)
* [enqueue/amqp-lib](https://github.com/php-enqueue/amqp-lib)
* [enqueue/amqp-bunny](https://github.com/php-enqueue/amqp-bunny)
* [enqueue/pheanstalk](https://github.com/php-enqueue/pheanstalk)
* [enqueue/stomp](https://github.com/php-enqueue/stomp)
* [enqueue/sqs](https://github.com/php-enqueue/sqs)
* [enqueue/rdkafka](https://github.com/php-enqueue/rdkafka)
* [enqueue/redis](https://github.com/php-enqueue/redis)
* [enqueue/gearman](https://github.com/php-enqueue/gearman)
* [enqueue/dbal](https://github.com/php-enqueue/dbal)
* [enqueue/fs](https://github.com/php-enqueue/fs)
* [enqueue/null](https://github.com/php-enqueue/null).
* [zogwarg/enqueue-elasticsearch](https://github.com/zogwarg/enqueue-elasticsearch)

It also provide an extended driver version for amqp interop compatible transports. It supports some amqp specific features such as queue declaration and message couting. 

TODO:

* [x] add tests
* [x] add doc

I welcome everyone who is interested in MQs to join the queue interop group

### CODE REVIEW

**Comment by acrobat** [comment]
Thanks for the PR @makasim!

@sagikazarmark should we still merge this or have new drivers directly in a separate repo like we are going to do in #256? Or still add it here and move the packages out all in one time?

**Comment by makasim** [comment]
Is it going to be merged? I need kind of approval to add tests and doc.

File: src/Driver/InteropDriver.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        if (false == array_key_exists($queueName, $this->consumers)) {
+            $queue = $this->context->createQueue($queueName);
+
+            $this->consumers[$queueName] = $this->context->createConsumer($queue);
+        }
+
+        return $this->consumers[$queueName];
+    }
+}
```
Comment (sagikazarmark) [inline src/Driver/InteropDriver.php]: Missing newline

File: src/Driver/InteropDriver.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
+<?php
+namespace Bernard\Driver;
+
+use Bernard\Driver;
+use Interop\Queue\PsrConsumer;
+use Interop\Queue\PsrContext;
+
+class InteropDriver implements Driver
```
Comment (sagikazarmark) [inline src/Driver/InteropDriver.php]: Should be final.

**Comment by sagikazarmark** [comment]
I would be more comfortable with a separate repo.

@henrikbjorn can you please grant @acrobat and me access to creating new repos. I would like to start separating drivers in a few weeks.

**Comment by makasim** [comment]
Could you please create a repo for me too so I  can move stuff there as well.

**Comment by kynx** [comment]
If someone's going to be separating out drivers can you have a look at https://github.com/bernardphp/OpenCloud? I did a pr for an opencloud driver ages ago- happy to pick it up again if there's any feedback.

**Comment by henrikbjorn** [comment]
I have granted the permissions.

man. 24. jul. 2017 kl. 20.50 skrev kynx <notifications@github.com>:

**Comment by henrikbjorn** [comment]
@kynx ill merge the opencloud one and feel free to pick it up again to match the new master,

**Comment by acrobat** [comment]
Yes @kynx will merge that pr! We will start soon with moving the driver to separate repo's and then we will also pick up your and this PR!

**Comment by sagikazarmark** [comment]
@henrikbjorn Thanks!

**Comment by sagikazarmark** [comment]
@henrikbjorn can you also please add me to a package on packagist as well?

> The vendor is already taken by someone else. You may ask them to add your package and give you maintainership access. If they add you as a maintainer on any package in that vendor namespace, you will then be able to add new packages in that namespace. The packages already in that vendor namespace can be found at bernard

**Comment by henrikbjorn** [comment]
What is your Packagist username?

**Comment by sagikazarmark** [comment]
mark.sagikazar

**Comment by henrikbjorn** [comment]
Done

**Comment by acrobat** [comment]
Thanks @henrikbjorn!

**Comment by sagikazarmark** [comment]
Thanks

**Comment by makasim** [comment]
Added AmqpInteropDriver which extends the functionality of InteropDriver with AMQP specific features. It could be used with php-amqplib, amqp-ext and bunny libs.

**Comment by makasim** [comment]
@sagikazarmark regarding separate repository. Isn't the code still merged to this repository and later it is split to the separate repository. If so, I don't see a reason why should we wait till it happens.

**Comment by makasim** [comment]
Added tests and docs. Ready for review.

**Comment by sagikazarmark** [comment]
> Isn't the code still merged to this repository and later it is split to the separate repository.

I am okay with merging code here (with a bit of squashing looking at the history), but I would like to avoid more releases before the driver refactor.

**Comment by makasim** [comment]
@sagikazarmark rebased&squashed

**Comment by makasim** [comment]
@sagikazarmark could you please take a look at this?

**Comment by acrobat** [comment]
I've took some time off and a bit of vacation in the last weeks, but I will make some time to review pr's in the coming days!

**Comment by makasim** [comment]
@acrobat just let me know if there is anything that should be done.

**Comment by sagikazarmark** [comment]
TBH I am not 100% sure why there is a separate driver for AMQP. That's the first thing that popped up after a quick review. Can't see any explanation in the docs either.

Other than that it's quite a PR, so I need some more time to review it.

**Comment by makasim** [comment]
@sagikazarmark [amqp interop](https://github.com/queue-interop/amqp-interop) is an exnteded version of [queue interop](https://github.com/queue-interop/queue-interop) and it defines AMQP specific things such as create\remove a queue, ability to get message count and so on. 

Obviously that any amqp interop compatible transport could be used with `InteropDriver` driver.


The amqp interop currently supports [three major amqp clients](https://github.com/queue-interop/queue-interop#compatible-projects-1) available in PHP: phpamqp-lib, amqp-ext and bunny and any of them will work with `AmqpInteropDriver`.

**Comment by makasim** [comment]
Worth to mention that I am going to provide support for  these interop drivers.

**Comment by sagikazarmark** [comment]
Thanks for the explanation, I thought something similar. Just for me to understand: the AmqpInterop driver contains some extra logic for AMQP implementations which are not supported by the plain Interop driver, but any AMQP implementation should work with the normal Interop driver as well.

Is that correct?

I wonder if it would make sense to extend the interop driver instead of passing it to the AmqpInterop driver. From the above explanation it seems it does.

**Comment by makasim** [comment]
> Is that correct?

you are absolutely right 

> I wonder if it would make sense to extend the interop driver instead of passing it to the AmqpInterop driver. 

we can have one single driver and some `if isntanceof AmqpInterop` inside to enhance amqp interop features. Let me know if it is more preferred way that the current one.

**Comment by sagikazarmark** [comment]
> we can have one single driver and some if isntanceof AmqpInterop inside to enhance amqp interop features. Let me know if it is more preferred way that the current one.

Well, if we can say that it's unlikely that over time we have to fill the Interop driver with tons of ifs to support all kinds of interop variants then that could work for me as well. Would make things simpler for sure.

**Comment by makasim** [comment]
I don't expect anything to be added over time.

**Comment by makasim** [comment]
@sagikazarmark merged everything in InteropDriver.

File: src/Driver/InteropDriver.php [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        $queue = $context->createQueue($queueName);
+        $queue->addFlag(AmqpQueue::FLAG_DURABLE);
+
+        return $queue;
+    }
+
+    /**
+     * @return PsrContext
+     */
+    public function getContext()
```
Comment (sagikazarmark) [inline src/Driver/InteropDriver.php]: Drivers are an internal thing to Bernard. Is there any reason to expose the context?

Comment (makasim) [inline src/Driver/InteropDriver.php]: I see. The method has been removed.

**Comment by makasim** [comment]
@sagikazarmark @acrobat a kind remainder. 

Sorry If I am insistent on this one, I don't want to put this into the pending tray

**Comment by sagikazarmark** [comment]
No problem, keep pushing it. I will give it a last shot today.

**Comment by sagikazarmark** [comment]
Thanks @makasim

**Comment by makasim** [comment]
Thank you too. Would you like to join queue interop group? I would really happy to see you there.

**Comment by acrobat** [comment]
:+1: thanks @makasim!

**Pull request merged**